internal/worker/charmrevision/worker.go

```diff
@@ -66,63 +138,528 @@ func (config Config) Validate() error {
+
+	origin := app.Origin
+	risk, err := encodeRisk(origin.Channel.Risk)
+	if err != nil {
+		return charmhubID{}, internalerrors.Errorf("encoding channel risk: %w", err)
+	}
+
+	channel, err := charm.MakeChannel(origin.Channel.Track, risk, origin.Channel.Branch)
+	if err != nil {
+		// It's actually impossible to get to the error here, as we've encoded
+		// the risk above. Yet, it's better to be safe.
+		return charmhubID{}, internalerrors.Errorf("making channel: %w", err)
+	}
+
+	arch, err := encodeArchitecture(origin.Platform.Architecture)
+	if err != nil {
+		return charmhubID{}, internalerrors.Errorf("encoding architecture: %w", err)
+	}
+
+	osType, err := encodeOSType(origin.Platform.OSType)
+	if err != nil {
+		return charmhubID{}, internalerrors.Errorf("encoding os type: %w", err)
+	}
+
+	return charmhubID{
+		id:          origin.ID,
+		revision:    origin.Revision,
+		channel:     channel.String(),
+		osType:      osType,
+		osChannel:   origin.Platform.Channel,
+		arch:        arch,
+		instanceKey: charmhub.CreateInstanceKey(appTag, modelTag),
+	}, nil
 }
 
-// Wait is part of the worker.Worker interface.
-func (ruw *revisionUpdateWorker) Wait() error {
-	return ruw.tomb.Wait()
+func encodeArchitecture(a architecture.Architecture) (string, error) {
+	switch a {
+	case architecture.AMD64:
+		return arch.AMD64, nil
+	case architecture.ARM64:
+		return arch.ARM64, nil
+	case architecture.PPC64EL:
+		return arch.PPC64EL, nil
+	case architecture.S390X:
+		return arch.S390X, nil
+	case architecture.RISCV64:
+		return arch.RISCV64, nil
+	default:
+		return "", internalerrors.Errorf("unsupported architecture %v", a)
+	}
+}
+
+func encodeOSType(t application.OSType) (string, error) {
+	switch t {
+	case application.Ubuntu:
+		return strings.ToLower(ostype.Ubuntu.String()), nil
+	default:
+		return "", internalerrors.Errorf("unsupported OS type %v", t)
+	}
+}
+
+func encodeRisk(r application.ChannelRisk) (string, error) {
+	switch r {
+	case application.RiskStable:
+		return internalcharm.Stable.String(), nil
+	case application.RiskCandidate:
+		return internalcharm.Candidate.String(), nil
+	case application.RiskBeta:
+		return internalcharm.Beta.String(), nil
+	case application.RiskEdge:
+		return internalcharm.Edge.String(), nil
+	default:
+		return "", internalerrors.Errorf("unsupported risk %v", r)
+	}
+}
+
+type appInfo struct {
+	id       string
+	charmURL *charm.URL
+}
+
+// charmhubID holds identifying information for several charms for a
+// charmhubLatestCharmInfo call.
+type charmhubID struct {
+	id        string
+	revision  int
+	channel   string
+	osType    string
+	osChannel string
+	arch      string
+	metrics   map[charmmetrics.MetricValueKey]string
+	// instanceKey is a unique string associated with the application. To assist
+	// with keeping KPI data in charmhub. It must be the same for every charmhub
+	// Refresh action related to an application.
+	instanceKey string
+}
+
+type latestCharmInfo struct {
+	url       *charm.URL
```

Comment:
I see that this is just copied over from the facade, but it would be really good if we could drop the dependency on `charm.URL` here. We don't use it anywhere here _yet_

Reply:
Interesting, let me think about that.

Reply:
I'm going to tackle this while setting the value in the database. I think we can drop charm URL completely.